### PR TITLE
Fix comparison for CHAR to take into account padding

### PR DESCRIPTION
**Pull request by maciejgrzybek**
Previous implementation of CHAR comparison didn't work properly with control chars (e.g. '\0').

---

Internal TD review: https://github.com/Teradata/presto/pull/363
Can be relevant: https://github.com/prestodb/presto/issues/2295

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/type/TestCharOperators.java [hunk 1]
```diff
@@ -29,6 +29,9 @@ public void testEqual()
         assertFunction("cast('foo' as char(3)) = cast('foo' as char(3))", BOOLEAN, true);
         assertFunction("cast('foo' as char(3)) = cast('bar' as char(3))", BOOLEAN, false);
         assertFunction("cast('bar' as char(3)) = cast('foo' as char(3))", BOOLEAN, false);
+
+        assertFunction("cast('a' as char(2)) = cast('a ' as char(2))", BOOLEAN, true);
+        assertFunction("cast('a ' as char(2)) = cast('a' as char(2))", BOOLEAN, true);
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/type/TestCharOperators.java]: add test cases:
```
        assertFunction("cast('a' as char(3)) <> cast('a' as char(2))", BOOLEAN, true);
        assertFunction("cast('' as char(3)) <> cast('' as char(2))", BOOLEAN, true);
        assertFunction("cast('' as char(2)) <> cast('' as char(2))", BOOLEAN, false);
```

for all operators

**Comment by sopel39** [comment]
Add test line to `TestTupleDomainOrcPredicate#testChars()`
```
        assertEquals(getDomain(CHAR, 10, stringColumnStats(10L, "\0 ", " ")), create(ValueSet.ofRanges(range(CHAR, utf8Slice("\0"), true, utf8Slice(""), true)), false));
```

**Comment by maciejgrzybek** [comment]
@sopel39 thanks for pointing me out to additional tests. Added in de53449.

**Comment by maciejgrzybek** [comment]
Rebased on master to resolve conflicts with `TestHiveDistributedQueries`.

File: presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java [hunk 2]
```diff
@@ -94,7 +95,7 @@ public boolean equalTo(Block leftBlock, int leftPosition, Block rightBlock, int
         if (leftLength != rightLength) {
             return false;
         }
-        return leftBlock.equals(leftPosition, 0, rightBlock, rightPosition, 0, leftLength);
+        return compareTo(leftBlock, leftPosition, rightBlock, rightPosition) == 0;
```
Comment (cberner) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java]: Why did you need to change the implementation of `equalTo`, and does `hash` need to change?

**Comment by cberner** [comment]
The product tests are failing, and I'm not sure why. Could you take a look?

**Comment by maciejgrzybek** [comment]
@cberner The problem was on master. Unrelated to this PR.
Rebased to the most recent master (where it's fixed already) made this PR pass on Travis.

Now it's ready to go. :)
